core/src/main/java/hudson/tasks/BuildStepCompatibilityLayer.java

```diff
@@ -70,16 +70,20 @@ public boolean prebuild(AbstractBuild<?,?> build, BuildListener listener) {
     @Override
     public boolean perform(AbstractBuild<?,?> build, Launcher launcher, BuildListener listener) throws InterruptedException, IOException {
         if (this instanceof SimpleBuildStep) {
+            final SimpleBuildStep sbs = ((SimpleBuildStep) this);
             // delegate to the overloaded version defined in SimpleBuildStep
             FilePath workspace = build.getWorkspace();
-            if (workspace == null) {
+            if (workspace == null && sbs.requiresWorkspace()) {
```

Comment:
Note that this would be a pathological condition: the build is almost certainly broken anyway.